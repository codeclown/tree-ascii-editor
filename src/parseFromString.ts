import { LineNode } from './types';

const unixStyle = /^(   |│   |├── |└── )+/;
const macStyle = /^(   |\|   |\|-- |`-- )+/;

export default function parseFromString(text: string): LineNode[] {
  const lines = text.split(/\n+/g).filter((line) => !line.match(/^\s*$/));
  return lines.map((line) => {
    const segments = line.match(unixStyle) || line.match(macStyle);
    const level = segments ? Math.ceil(segments[0].length / 4) : 0;
    const text = line.slice(level * 4);
    return {
      type: 'line',
      level,
      children: [{ text }],
    };
  });
}
