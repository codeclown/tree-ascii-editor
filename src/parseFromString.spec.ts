import parseFromString from './parseFromString';

describe('parseFromString', () => {
  it('parses nothing', () => {
    expect(parseFromString('')).toEqual([]);
    expect(parseFromString('\n')).toEqual([]);
    expect(parseFromString(' ')).toEqual([]);
    expect(parseFromString('  ')).toEqual([]);
  });

  it('parses unix tree output', () => {
    expect(
      parseFromString(
        `
path/to/folder/
├── a-first.html
├── b-second.html
├── subfolder
│   ├── readme.html
│   ├── code.cpp
│   ├── sub
│   │   └── file.html
│   └── code.h
└── z-last-file.html
        `.trim()
      )
    ).toMatchSnapshot();

    expect(
      parseFromString(
        `
path/to/folder/
└── subfolder
    └── sub
        └── file.html
        `.trim()
      )
    ).toMatchSnapshot();
  });

  it('parses mac tree output', () => {
    expect(
      parseFromString(
        `
path/to/folder/
|-- a-first.html
|-- b-second.html
|-- subfolder
|   |-- readme.html
|   |-- code.cpp
|   |-- sub
|   |   \`-- file.html
|   \`-- code.h
\`-- z-last-file.html
        `.trim()
      )
    ).toMatchSnapshot();

    expect(
      parseFromString(
        `
path/to/folder/
\`-- subfolder
    \`-- sub
        \`-- file.html
        `.trim()
      )
    ).toMatchSnapshot();
  });
});
