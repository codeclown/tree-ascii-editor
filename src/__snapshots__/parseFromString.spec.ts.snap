// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFromString parses mac tree output 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "path/to/folder/",
      },
    ],
    "level": 0,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "a-first.html",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "b-second.html",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "subfolder",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "readme.html",
      },
    ],
    "level": 2,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "code.cpp",
      },
    ],
    "level": 2,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "sub",
      },
    ],
    "level": 2,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "file.html",
      },
    ],
    "level": 3,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "code.h",
      },
    ],
    "level": 2,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "z-last-file.html",
      },
    ],
    "level": 1,
    "type": "line",
  },
]
`;

exports[`parseFromString parses mac tree output 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "path/to/folder/",
      },
    ],
    "level": 0,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "subfolder",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "\`-- sub",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "\`-- file.html",
      },
    ],
    "level": 2,
    "type": "line",
  },
]
`;

exports[`parseFromString parses unix tree output 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "path/to/folder/",
      },
    ],
    "level": 0,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "a-first.html",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "b-second.html",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "subfolder",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "readme.html",
      },
    ],
    "level": 2,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "code.cpp",
      },
    ],
    "level": 2,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "sub",
      },
    ],
    "level": 2,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "file.html",
      },
    ],
    "level": 3,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "code.h",
      },
    ],
    "level": 2,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "z-last-file.html",
      },
    ],
    "level": 1,
    "type": "line",
  },
]
`;

exports[`parseFromString parses unix tree output 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "path/to/folder/",
      },
    ],
    "level": 0,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "subfolder",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "└── sub",
      },
    ],
    "level": 1,
    "type": "line",
  },
  Object {
    "children": Array [
      Object {
        "text": "└── file.html",
      },
    ],
    "level": 2,
    "type": "line",
  },
]
`;
